samplers.go

```diff
@@ -23,6 +27,97 @@ type DDMetric struct {
 	Interval   int32         `json:"interval,omitempty"`
 }
 
+type tsvField int
+
+const (
+	// the order in which these appear determines the
+	// order of the fields in the resultant TSV
+	tsvName tsvField = iota
+	tsvTags
+	tsvMetricType
+	tsvHostname
+	tsvDeviceName
+	tsvInterval
+
+	tsvTimestamp
+	tsvValue
+)
+
+var tsvSchema = [...]string{
+	tsvName:       "tsvName",
+	tsvTags:       "tsvTags",
+	tsvMetricType: "tsvMetricType",
+	tsvHostname:   "tsvHostname",
+	tsvDeviceName: "tsvDeviceName",
+	tsvInterval:   "tsvInterval",
+	tsvTimestamp:  "tsvTimestamp",
+	tsvValue:      "tsvValue",
+}
+
+// String returns the field name.
+// eg tsvName.String() returns "Name"
+func (f tsvField) String() string {
+	return fmt.Sprintf(strings.Replace(tsvSchema[f], "tsv", "", 1))
+}
+
+// each key in tsvMapping is guaranteed to have a unique value
+var tsvMapping = map[string]int{}
+
+func init() {
+	for i, field := range tsvSchema {
+		tsvMapping[field] = i
+	}
+}
+
+// encodeCSV generates a newline-terminated CSV row that describes
+// the data represented by the DDMetric.
+// The caller is responsible for setting w.Comma as the appropriate delimiter.
+// For performance, encodeCSV does not flush after every call; the caller is
+// expected to flush at the end of the operation cycle
+func (d DDMetric) encodeCSV(w *csv.Writer) error {
+
+	timestamp := d.Value[0][0]
+	value := strconv.FormatFloat(d.Value[0][1], 'f', -1, 64)
+	interval := strconv.Itoa(int(d.Interval))
+
+	// TODO(aditya) some better error handling for this
+	// to guarantee that the result is proper JSON
+	tags := "[" + strings.Join(d.Tags, ",") + "]"
```

Comment:
I assume you'll fix this later, but to be clear, what we want here is a JSON object `{"foo":"bar", "baz":"qux"}`, rather than a JSON array `["foo:bar","baz:qux"]`